backend/pyutils/csv_to_model.py

```diff
@@ -1,42 +1,60 @@
+""" Convert Immunisation CSV data to Immunization Model """
 import csv
-from backend.models.immunization import ImmunizationModel
-from datetime import datetime
-
-
-def parse_iso8601_datetime(iso_datetime_str):
-    try:
-        # Parse ISO 8601 datetime string with timezone offset
-        dt = datetime.fromisoformat(iso_datetime_str)
-
-        # Convert the parsed datetime to the system's local timezone
-        dt = dt.astimezone()
-
-        return dt
-    except ValueError:
-        return None
-
-
-# Define a function to read and transform the CSV file
-def read_csv_to_immunizations(csv_data):
-    immunizations = [] = []
-
-    csv_reader = csv.DictReader(csv_data.splitlines())
-    for row in csv_reader:
-        # Parse datetime fields if needed
-        if row["PERSON_DOB"]:
-            row["PERSON_DOB"] = datetime.strptime(row["PERSON_DOB"], "%Y-%m-%d").date()
-        if row["DATE_AND_TIME"]:
-            row["DATE_AND_TIME"] = parse_iso8601_datetime(row["DATE_AND_TIME"])
-        if row["RECORDED_DATE"]:
-            row["RECORDED_DATE"] = datetime.strptime(
-                row["RECORDED_DATE"], "%Y-%m-%d"
-            ).date()
-        if row["EXPIRY_DATE"]:
-            row["EXPIRY_DATE"] = datetime.strptime(
-                row["EXPIRY_DATE"], "%Y-%m-%d"
-            ).date()
-        immunization = ImmunizationModel(**row)
-        if immunization is not None:
-            immunizations.append(immunization)
-
-    return immunizations
+import os
+import sys
+from typing import Union
+from pydantic import ValidationError
+
+sys.path.append(f"{os.path.dirname(os.path.abspath(__file__))}/../")
+
+
+from models.immunization import ImmunizationModel
+from models.failures import ImmunizationErrorModel
+
+
+def generate_failed_record(row: dict, reasons: str) -> ImmunizationErrorModel:
+    """Format an exception into a record"""
+    failed_record = {
+        "nhs_number": row.get("NHS_NUMBER", None),
+        "unique_id": row.get("UNIQUE_ID", None),
+        "failure_reasons": reasons,
+    }
+
+    immunisation_failure = ImmunizationErrorModel(**failed_record)
+
+    return immunisation_failure
+
+
+def read_csv_to_model(
```

Comment:
This function shouldn't expect a file. This code should perform under the assumption that it receives the content as string. You can read a file during test and pass it to this function. If this function assumes file then, we can't use it in the actual implementation because the content comes from S3 not reading a file.